src/allotropy/parsers/beckman_pharmspec/beckman_pharmspec_parser.py

```diff
@@ -259,4 +263,79 @@ def _setup_model(self, df: pd.DataFrame, file_name: str) -> Model:
                 calculated_data_aggregate_document=calc_agg_doc,
             ),
         )
-        return model
+
+    def _get_distribution_id_for_run_and_particle_size(
+        self,
+        run_name: str,
+        particle_size: float,
+        measurement_doc_items: list[MeasurementDocumentItem],
+    ) -> Optional[Union[str, TStringValueItem]]:
+        item = next(
+            x for x in measurement_doc_items if x.measurement_identifier == run_name
+        )
+        if (
+            item.processed_data_aggregate_document is not None
+            and item.processed_data_aggregate_document.processed_data_document
+            is not None
+        ):
+            for pdd in item.processed_data_aggregate_document.processed_data_document:
+                if (
+                    pdd.distribution_aggregate_document is not None
+                    and pdd.distribution_aggregate_document.distribution_document
+                    is not None
+                ):
+                    for dd in pdd.distribution_aggregate_document.distribution_document:
+                        for d in dd.distribution:
+                            if d.particle_size.value == particle_size:
+                                return d.distribution_identifier
+        return None
+
+    def _add_data_source_to_calculated_data(
+        self,
+        calc_agg_doc: TCalculatedDataAggregateDocument,
+        measurement_doc_items: list[MeasurementDocumentItem],
+    ) -> None:
+        if calc_agg_doc.calculated_data_document is not None:
+            for cdd in calc_agg_doc.calculated_data_document:
+                if (
+                    cdd.data_source_aggregate_document is not None
+                    and cdd.data_source_aggregate_document.data_source_document
+                    is not None
+                ):
+                    for dsd in cdd.data_source_aggregate_document.data_source_document:
+                        run_name, particle_size = str(dsd.data_source_identifier).split(
+                            "|"
+                        )
+                        distribution_id = (
+                            self._get_distribution_id_for_run_and_particle_size(
+                                run_name, float(particle_size), measurement_doc_items
+                            )
+                        )
+                        if distribution_id is not None:
+                            dsd.data_source_identifier = distribution_id
+
+    def _setup_model(self, df: pd.DataFrame, file_name: str) -> Model:
```

Comment:
I'm not going to block this PR on this, but a request.

This parser is getting complicated enough that it probably makes sense to break out this prep work into a set of structure classes.

Generally - we want the parser class to be responsible for calling out to other classes that read the data and using them to construct the model.

We skip the structure pattern when a parser is really simple, but this is going to the point where I think it should be factored out.

Reply:
Makes sense. I'll follow-up.